src/layer/vox/index.ts

```diff
@@ -0,0 +1,359 @@
+      let sum = 0;
+      for (
+        let globalDim = 0;
+        globalDim < Math.min(globalRank, 3);
+        ++globalDim
+      ) {
+        const layerDim = globalToRenderLayerDimensions[globalDim];
+        if (layerDim !== -1) {
+          sum +=
+            layerToChunkTransform[chunkDim + layerDim * (layerRank + 1)] *
+            globalNormal[globalDim];
+        }
+      }
+      voxelNormal[chunkDim] = sum;
+    }
+    vec3.normalize(voxelNormal, voxelNormal);
+    return voxelNormal;
+  }
+}
+
+export declare abstract class UserLayerWithVoxelEditing extends UserLayer {
+  isEditable: WatchableValue<boolean>;
+
+  voxBrushRadius: TrackableValue<number>;
+  voxEraseMode: TrackableBoolean;
+  voxBrushShape: TrackableEnum<BrushShape>;
+  voxFloodMaxVoxels: TrackableValue<number>;
+  paintValue: TrackableValue<bigint>;
+
+  editingContexts: Map<LoadedDataSubsource, VoxelEditingContext>;
+
+  abstract _createVoxelRenderLayer(
+    source: MultiscaleVolumeChunkSource,
+    transform: WatchableValueInterface<RenderLayerTransformOrError>,
+  ): ImageRenderLayer | SegmentationRenderLayer;
+  abstract getVoxelPaintValue(erase: boolean): bigint;
+  abstract setVoxelPaintValue(value: bigint): void;
+
+  initializeVoxelEditingForSubsource(
+    loadedSubsource: LoadedDataSubsource,
+    renderlayer: SegmentationRenderLayer | ImageRenderLayer,
+  ): void;
+  deinitializeVoxelEditingForSubsource(
+    loadedSubsource: LoadedDataSubsource,
+  ): void;
+
+  getIdentitySliceViewSourceOptions(): SliceViewSourceOptions;
+  handleVoxAction(action: string, context: LayerActionContext): void;
+}
+
+export function UserLayerWithVoxelEditingMixin<
+  TBase extends { new (...args: any[]): UserLayerWithAnnotations },
+>(Base: TBase) {
+  abstract class C extends Base implements UserLayerWithVoxelEditing {
+    editingContexts = new Map<LoadedDataSubsource, VoxelEditingContext>();
+    isEditable = new WatchableValue<boolean>(false);
+    paintValue = new TrackableValue<bigint>(1n, (x) => parseUint64(x));
+
+    // Brush properties
+    voxBrushRadius = new TrackableValue<number>(3, verifyInt);
+    voxEraseMode = new TrackableBoolean(false);
+    voxBrushShape = new TrackableEnum(BrushShape, BrushShape.DISK);
+    voxFloodMaxVoxels = new TrackableValue<number>(10000, verifyFiniteFloat);
+
+    constructor(...args: any[]) {
+      super(...args);
+      this.registerDisposer(() => {
+        for (const context of this.editingContexts.values()) {
+          context.dispose();
+        }
+        this.editingContexts.clear();
+      });
+      this.voxBrushRadius.changed.add(this.specificationChanged.dispatch);
+      this.voxEraseMode.changed.add(this.specificationChanged.dispatch);
+      this.voxBrushShape.changed.add(this.specificationChanged.dispatch);
+      this.voxFloodMaxVoxels.changed.add(this.specificationChanged.dispatch);
+      this.paintValue.changed.add(this.specificationChanged.dispatch);
+      this.tabs.add("Draw", {
```

Comment:
You can use the `hidden` property to ensure this is only visible if there is a writable source.